secure/secure.sh

```diff
@@ -133,24 +139,35 @@ function krb5_configuration() {
     sed -i "/\[domain_realm\]/a\ \t${cross_realm_trust_kdc} = ${cross_realm_trust_realm}\n\t\.${DOMAIN} = ${REALM}" "/etc/krb5.conf"
   fi
   echo "Configuring krb5 to add local realm in /etc/krb5.conf."
-  sed -i "/\[realms\]/a\ \t${REALM} = {\n\t\tkdc = ${MASTER_FQDN}\n\t\tadmin_server = ${MASTER_FQDN}\n\t}" "/etc/krb5.conf"
+  additional_kdcs=""
+  if [[ "${HA_MODE}" == 1 ]]; then
+    for additional_master in "${ADDITIONAL_MASTERS[@]}"; do
+      additional_kdcs+="\t\tkdc = ${additional_master}.${DOMAIN}\n"
+    done
+  fi
+  sed -i "/\[realms\]/a\ \t${REALM} = {\n\t\tkdc = ${MASTER_FQDN}\n${additional_kdcs}\t\tadmin_server = ${MASTER_FQDN}\n\t}" "/etc/krb5.conf"
 }
 
-function create_root_user_principal() {
-  if [[ "${ROLE}" == 'Master' ]]; then
+function create_root_user_principal_and_ready_stash() {
+  if [[ "${FQDN}" == "${MASTER_FQDN}" ]]; then
     echo "Creating root principal root@${REALM}."
     echo -e "${root_principal_password}\n${root_principal_password}" | /usr/sbin/kadmin.local -q "addprinc root"
-    # write a file to gcs to signal to workers that Kerberos server is ready
+    if [[ "${HA_MODE}" == 1 ]]; then
+      echo "In HA mode, writing krb5 stash file to GCS: gs://${dataproc_bucket}/${CLUSTER_UUID}/stash"
```

Comment:
Rather than writing directly in the root of this bucket, consider using `gs://<bucket>/google-cloud-dataproc-metainfo/<cluster_uuid>/secure_init_action/` to match what Dataproc uses internally.

Reply:
done.